lib/stm32/l4/rcc.c

```diff
@@ -369,4 +454,136 @@ void rcc_set_msi_range_standby(uint32_t msi_range)
 	RCC_CSR = reg;
 }
 
+/** Enable PLL Output
+ *
+ *	- P (RCC_PLLCFGR_PLLPEN)
+ *	- Q (RCC_PLLCFGR_PLLQEN)
+ *	- R (RCC_PLLCFGR_PLLREN)
+ *
+ * @param pllout One or more of the definitions above
+ */
+void rcc_pll_output_enable(uint32_t pllout)
+{
+	RCC_PLLCFGR |= pllout;
+}
+
+/** Set clock source for 48MHz clock
+ *
+ * The 48 MHz clock is derived from one of the four following sources:
+ *	- main PLL VCO (RCC_CCIPR_CLK48SEL_PLL)
+ *	- PLLSAI1 VCO (RCC_CCIPR_CLK48SEL_PLLSAI1Q)
+ *	- MSI clock (RCC_CCIPR_CLK48SEL_MSI)
+ *	- HSI48 internal oscillator (RCC_CCIPR_CLK48SEL_HSI48)
+ *
+ * @param clksel One of the definitions above
+ */
+void rcc_set_clock48_source(uint32_t clksel)
+{
+	RCC_CCIPR &= ~(RCC_CCIPR_CLK48SEL_MASK << RCC_CCIPR_CLK48SEL_SHIFT);
+	RCC_CCIPR |= (clksel << RCC_CCIPR_CLK48SEL_SHIFT);
+}
+
+void rcc_clock_setup_hsi16(void)
```

Comment:
why isn't this just taking a clock structure too? this seems odd and out of place.  is this leftover from something else?

Reply:
Code seemed clearer and simpler that way. Other families have similar different functions for different clock sources. I am actually using it in bootloader code...

Reply:
which ones? you have right below one that uses hsi16 as the clock source and takes a structure?

Reply:
ok, but it uses PLL as sysclock source, now hsi raw.

L1: rcc_clock_setup_msi(), rcc_clock_setup_hsi(), rcc_clock_setup_pll()
F1: rcc_clock_setup_in_hsi_out_64mhz(), rcc_clock_setup_in_hsi_out_48mhz(), rcc_clock_setup_in_hsi_out_24mhz(), rcc_clock_setup_in_hse_8mhz_out_24mhz(), ... etc... just to name a few

Reply:
no, not just to name a few :)  F1 is _not_ the example to follow.  Your clock strcut should just include the source.